src/lib/redisConnector.ts

```diff
@@ -0,0 +1,42 @@
+import { type RedisClientType, createClient } from "redis";
+import { REDIS_URL } from "@src/config";
+
+export class RedisConnector {
+  private static instance: RedisConnector | undefined = undefined;
+  private static RETRY_MAX = 10;
+  private static RETRY_DELAY_STEP = 500; // milliseconds
+
+  public client: RedisClientType;
+
+  private constructor() {
+    this.client = createClient({
+      url: REDIS_URL,
+      socket: {
+        // Reconnect strategy with exponential backoff
+        reconnectStrategy: (retries: number): number | Error =>
+          retries < RedisConnector.RETRY_MAX
+            ? RedisConnector.RETRY_DELAY_STEP * retries
+            : new Error("Failed to connect to Redis"),
+      },
+    });
+    this.client
+      .on("error", (err: Error) => console.error("Redis client error:", err))
+      .on("ready", () => console.debug("Redis client ready!"))
+      .on("reconnecting", () => console.debug("Redis client reconnecting..."));
+  }
+
+  /**
+   * Uses the singleton promise pattern to initialize the RedisConnector class instance.
+   * This ensures that only one connection is attempted to the Redis server when the
+   * instance is first initialized, even if multiple concurrent calls are made.
+   * @returns {Promise<RedisConnector>}
+   */
+  // biome-ignore lint/suspicious/useAwait: Singleton Promise is resolved by the caller
+  public static async getInstance(): Promise<RedisConnector> {
+    if (RedisConnector.instance === undefined) {
+      RedisConnector.instance = new RedisConnector();
+      RedisConnector.instance.client.connect();
```

Comment:
So initially I had gone with that same approach but it meant that multiple concurrent calls to `getInstance()` when the instance was `undefined` caused multiple connection attempts and overwriting of the singleton instance.

My thinking for the `await` being required on the call to `getInstance()` rather than on the `connect()` was it would be blocking for all concurrent call.  Only the first call would cause a connection attempt, and once that promise resolved, all concurrent calls would resolve and use the same connection.  

It was based on [this post](https://www.jonmellman.com/posts/singleton-promises) talking about the singleton promise pattern but I'm also more than happy to adjust.

Reply:
I'd be more than happy to connect and walk through the code.  I've seen this pattern before and I've also used it on my branch to connect to the GCForms DB.
The getInstance() does need to return a promise but Biome is complaining because you're using an async function that does not include an await.  Best practices would ensure that if a promise needs to resolve before returning that it should be awaited.  In this case we want the `connect()` to resolve before returning the instance back to the caller.  Including the await will not overwrite the singleton.  Awaiting the `getInstance` does not block any other concurrent calls.  I'll DM you an CodeSandbox with a demo that shows using a Promise.All on several calls to `getInstance()`.

Reply:
Much appreciated for walking me through it in your sandbox - I'll get the change in!